core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -75,6 +77,17 @@
     public static final Setting<Settings> CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING =
         Setting.groupSetting(CLUSTER_ROUTING_EXCLUDE_GROUP_PREFIX + ".", IP_VALIDATOR, Property.Dynamic, Property.NodeScope);
 
+    /**
+     * The set of {@link RecoverySource.Type} values for which the
+     * {@link IndexMetaData#INDEX_ROUTING_INITIAL_RECOVERY_GROUP_SETTING} should apply.
+     * Note that we do not include the {@link RecoverySource.Type#SNAPSHOT} type here
+     * because if the snapshot is restored to a different cluster that does not contain
+     * the initial recovery node id, or to the same cluster where the initial recovery node
+     * id has been decommissioned, then the primary shards will never be allocated.
+     */
+    public static EnumSet<RecoverySource.Type> INITIAL_RECOVERY_TYPES =
```

Comment:
the problem is that `FilterAllocationDeciderTests` is not part of the same package, and moving it to the same package as `FilterAllocationDecider` breaks package visibility on the `AllocationService#reroute` protected method that the test depends on.